### PR TITLE
reduce precompile() failure severity to a warning

### PR DESCRIPTION
**Pull request by vtjnash**
Many users (including Base) are calling `@assert`, despite that this is
not what assert should be used to mark, for many reasons.

This happened to also reveal a small number of errors, so also detect
those (for fixing later).

Refs: https://github.com/JuliaLang/julia/commit/c0f9666d0b94b213c7ff9e64a7b4e5268aa0e18b#commitcomment-47782674

### CODE REVIEW

File: base/loading.jl [hunk 1]
```diff
@@ -1218,11 +1218,9 @@ function include_package_for_output(pkg::PkgId, input::String, depot_path::Vecto
     end
 end
 
-@assert precompile(include_package_for_output, (PkgId,String,Vector{String},Vector{String},Vector{String},typeof(_concrete_dependencies),Nothing))
```
Comment (timholy) [inline base/loading.jl:1221]: I'm not looking at the context here (in the middle of other things), but if memory serves it's actually an error if this doesn't get precompiled?

**Review by timholy — DISMISSED** [review]
Works for me, especially if the warnings help follow up on bugs.

File: base/loading.jl (hunk 1 above)
Comment (DilumAluthge) [reply base/loading.jl:1221]: We still shouldn't use `@assert` here. A future minor release of Julia is allowed to disable assertions by default. If you really need to throw an error here, you should do it explicitly.

Comment (DilumAluthge) [reply base/loading.jl:1221]: E.g. 
```
result = precompile(include_package_for_output, (PkgId,String,Vector{String},Vector{String},Vector{String},typeof(_concrete_dependencies),Nothing))
if !result
    msg = "failed to precompile"
    throw(ErrorException(msg))
end
```

Or something like that.

That will ensure that the error is always thrown, even if assertions are disabled by default in a future release of Julia.

**Comment by timholy** [comment]
Now that I have a bit more time, I'd like to suggest a different approach. Let me first acknowledge that using `@assert` this way was an abuse: it was a lazy choice that unnecessarily turned the detection of a few wisps of smoke into a 3-alarm fire. And @DilumAluthge's point about potential changes in `@assert` are spot-on. Writing and using `@warnfail` would be a much better way of handling this need.

But there are a few cases where precompilation *must* succeed for something else to work. I'm not certain about the example of `include_package_for_output`, but it's certainly true for https://github.com/timholy/SnoopCompile.jl/blob/d8feaa1711d548f3cff0e1462778abc8a7ac98f4/SnoopCompileCore/src/snoopi.jl#L89-L104. When precompilation fails it might be a lot harder to diagnose through the later consequences, so in such cases I do think it's much better to take a stop-the-presses approach and deal with the problem immediately.

So my proposal is that I implement `@warnfail` in SnoopCompile, change the docs, and submit PRs fixing it in Julia's `generate_precompiles.jl` and various packages. Meanwhile we leave the implementation of `precompile` the way it is, other than fixing any immediate bugs. And then after getting 1.6 out the door we elevate the priority of fixing the more difficult issues that have been filed about precompilation's mysterious failures.

**Comment by timholy** [comment]
Emergency over :smile:. See https://github.com/JuliaLang/julia/commit/c0f9666d0b94b213c7ff9e64a7b4e5268aa0e18b#commitcomment-47837689

**Pull request merged**